Add example how lints can be feature gated

xFrednet opened the pull request:
Just a simple addition, like the title says.

I hope that everyone reading this is having a great day. Keep up the good work :upside_down_face:

camelid commented on src/diagnostics.md:
```diff
@@ -594,6 +594,20 @@ declare_lint! {
+    pub SOME_LINT_NAME,
+    Warn,
+    "a new and useful, but feature gated lint",
+    @feature_gate = rustc_span::sym::feature_name;
```
`sym` is already imported.

```suggestion
    @feature_gate = sym::feature_name;
```

camelid commented on src/diagnostics.md:
```diff
@@ -594,6 +594,20 @@ declare_lint! {
+### Feature-gated lints
+
+Lints belonging to a feature should only be usable if the feature is enabled in the
+crate. To support this, lint declarations can contain a feature gate like this:
```
```suggestion
crate. To support this, lint declarations can contain a feature gate like so:
```

xFrednet replied:
`this` is one of the words I use way to often, nice improvement :upside_down_face:

camelid reviewed (approved):
Thanks for this PR! Just two suggestions and then LGTM.

One other thing: I think sometimes lints belonging to a feature are feature-gated but in the past they were not. Is the compiler/lang team's policy to always feature-gate lints belonging to a feature now? We don't really have to decide now since lints can always be un-feature-gated later if necessary, but I wanted to bring it up.

xFrednet commented:
Thank you for the review, I've added your suggestions and squashed the commits.

> Is the compiler/lang team's policy to always feature-gate lints belonging to a feature now? We don't really have to decide now since lints can always be un-feature-gated later if necessary, but I wanted to bring it up.

I don't know of any policies in that regard, it most likely depends on the lint. For instance, I'm currently working on the implementation of the `expect` attribute that emits a lint under specific circumstances. There a feature gate should be used IMO, as the lint can and should only be emitted if the feature is enabled. With other lints, it will most likely be a different story :upside_down_face:

xFrednet commented:
@rustbot ready

xFrednet commented:
@rustbot label -aiting-on-author
